plugins/trace_api_plugin/test/test_trace_file.cpp

```diff
@@ -0,0 +1,405 @@
+         "b000000000000000000000000000000000000000000000000000000000000001"_h, 5, 0
+      } };
+      const metadata_log_entry le1 { lib_entry_v0 { 4 } };
+      const metadata_log_entry be2 { block_entry_v0 {
+         "b000000000000000000000000000000000000000000000000000000000000002"_h, 7, 0
+      } };
+      const metadata_log_entry le2 { lib_entry_v0 { 5 } };
+
+   };
+
+   class vslice_datastream;
+
+   struct vslice {
+      enum mode { read_mode, write_mode};
+      vslice(mode m = write_mode) : _mode(m) {}
+      long tellp() const {
+         return _pos;
+      }
+
+      void seek( long loc ) {
+         if (_mode == read_mode) {
+            if (loc > _buffer.size()) {
+               throw std::ios_base::failure( "read vslice unable to seek to: " + std::to_string(loc) + ", end is at: " + std::to_string(_buffer.size()));
+            }
+         }
+         _pos = loc;
+      }
+
+      void seek_end( long loc ) {
+         _pos = _buffer.size();
+      }
+
+      void read( char* d, size_t n ) {
+         if( _pos + n > _buffer.size() ) {
+            throw std::ios_base::failure( "vslice unable to read " + std::to_string( n ) + " bytes; only can read " + std::to_string( _buffer.size() - _pos ) );
+         }
+         std::memcpy( d, _buffer.data() + _pos, n);
+         _pos += n;
+      }
+
+      void write( const char* d, size_t n ) {
+         if (_mode == read_mode) {
+            throw std::ios_base::failure( "read vslice should not have write called" );
+         }
+         if (_buffer.size() < _pos + n) {
+            _buffer.resize(_pos + n);
+         }
+         std::memcpy( _buffer.data() + _pos, d, n);
+         _pos += n;
+      }
+
+      void flush() {
+         _flush = true;
+      }
+
+      void sync() {
+         _sync = true;
+      }
+
+      vslice_datastream create_datastream();
+
+      std::vector<char> _buffer;
+      mode _mode = write_mode;
+      long _pos = 0l;
+      bool _flush = false;
+      bool _sync = false;
+   };
+
+   class vslice_datastream {
+   public:
+      explicit vslice_datastream( vslice& vs ) : _vs(vs) {}
+
+      void skip( size_t s ) {
+         std::vector<char> d( s );
+         read( &d[0], s );
+      }
+
+      bool read( char* d, size_t s ) {
+         _vs.read( d, s );
+         return true;
+      }
+
+      bool get( unsigned char& c ) { return get( *(char*)&c ); }
+
+      bool get( char& c ) { return read(&c, 1); }
+
+   private:
+      vslice& _vs;
+   };
+
+   inline vslice_datastream vslice::create_datastream() {
+      return vslice_datastream(*this);
+   }
+}
+
+BOOST_AUTO_TEST_SUITE(slice_tests)
+   BOOST_FIXTURE_TEST_CASE(write_data_trace, test_fixture)
+   {
+      vslice vs;
+      const auto offset = append_data_log( bt, vs );
+      BOOST_REQUIRE_EQUAL(offset,0);
+
+      const auto expected_offset = vs._pos;
+      const auto offset2 = append_data_log( bt2, vs );
+      BOOST_REQUIRE_EQUAL(offset2, expected_offset);
```

Comment:
Is this a necessary invariant to test?  If, for example, in the future we introduce some alignment/padding or internal data in between entries this test will break but I don't think the overall system will be broken.  As long as it returns an offset which can be used to retrieve the exact data (as you test on the lines below)

Reply:
not necessary, mostly providing training wheels at the outset, can remove now